sources/Renderer/OpenGL/Buffer/GLBuffer.cpp

```diff
@@ -291,54 +306,106 @@ void GLBuffer::CopyBufferSubData(const GLBuffer& readBuffer, GLintptr readOffset
 void* GLBuffer::MapBuffer(GLenum access)
 {
-    #if LLGL_GLEXT_DIRECT_STATE_ACCESS
-    if (HasExtension(GLExt::ARB_direct_state_access))
-    {
-        return glMapNamedBuffer(GetID(), access);
-    }
+    #ifdef GL_ARB_buffer_storage
+    if (storageFlags_ & GL_MAP_PERSISTENT_BIT)
+        {
+            if (mapped_ == nullptr)
+            {
+                GLbitfield flags = storageFlags_ & (GL_MAP_READ_BIT | GL_MAP_WRITE_BIT | GL_MAP_PERSISTENT_BIT | GL_MAP_COHERENT_BIT);
+                #if LLGL_GLEXT_DIRECT_STATE_ACCESS
+                if (HasExtension(GLExt::ARB_direct_state_access))
+                {
+                    mapped_ = glMapNamedBufferRange(GetID(), 0, bufferSize_, flags);
+                }
+                else
+                #endif // /LLGL_GLEXT_DIRECT_STATE_ACCESS
+                {
+                    GLStateManager::Get().BindGLBuffer(*this);
+                    mapped_ = glMapBufferRange(GetGLTarget(), 0, bufferSize_, flags);
+                }
+            }
+            return mapped_;
+        }    
     else
-    #endif // /LLGL_GLEXT_DIRECT_STATE_ACCESS
+    #endif // /GL_ARB_buffer_storage
     {
-        GLStateManager::Get().BindGLBuffer(*this);
-        return GLProfile::MapBuffer(GetGLTarget(), access);
+        #if LLGL_GLEXT_DIRECT_STATE_ACCESS
+        if (HasExtension(GLExt::ARB_direct_state_access))
+        {
+            return glMapNamedBuffer(GetID(), access);
+        }
+        else
+        #endif // /LLGL_GLEXT_DIRECT_STATE_ACCESS
+        {
+            GLStateManager::Get().BindGLBuffer(*this);
+            return GLProfile::MapBuffer(GetGLTarget(), access);
+        }
     }
 }
 
 void* GLBuffer::MapBufferRange(GLintptr offset, GLsizeiptr length, GLbitfield access)
 {
-    #if LLGL_GLEXT_DIRECT_STATE_ACCESS
-    if (HasExtension(GLExt::ARB_direct_state_access))
-    {
-        return glMapNamedBufferRange(GetID(), offset, length, access);
-    }
-    else
-    #endif // /LLGL_GLEXT_DIRECT_STATE_ACCESS
-    #ifdef GL_ARB_map_buffer_range
-    if (HasExtension(GLExt::ARB_map_buffer_range))
+    #ifdef GL_ARB_buffer_storage
+    if (storageFlags_ & GL_MAP_PERSISTENT_BIT)
     {
-        GLStateManager::Get().BindGLBuffer(*this);
-        return glMapBufferRange(GetGLTarget(), offset, length, access);
+        if (mapped_ == nullptr)
+        {
+            GLbitfield flags = storageFlags_ & (GL_MAP_READ_BIT | GL_MAP_WRITE_BIT | GL_MAP_PERSISTENT_BIT | GL_MAP_COHERENT_BIT);
+            #if LLGL_GLEXT_DIRECT_STATE_ACCESS
+            if (HasExtension(GLExt::ARB_direct_state_access))
+            {
+                mapped_ = glMapNamedBufferRange(GetID(), 0, bufferSize_, flags);
+            }
+            else
+            #endif // /LLGL_GLEXT_DIRECT_STATE_ACCESS
+            {
+                GLStateManager::Get().BindGLBuffer(*this);
+                mapped_ = glMapBufferRange(GetGLTarget(), 0, bufferSize_, flags);
+            }
+        }
+        return (char*)mapped_ + offset;
```

Comment:
Please don't use C-style casts, they can unintentionally cast incompatible types, cast constness away, and silence compiler warnings. This can be a `static_cast` since we're casting from `void*` to something else.